enterprise/frontend/src/embedding-sdk/reducer.ts

```diff
@@ -0,0 +1,66 @@
+import { createAsyncThunk, createReducer } from "@reduxjs/toolkit";
+import type {
+  EmbeddingSessionTokenState,
+  EnterpriseState,
+} from "embedding-sdk/types";
+
+const initialState: EmbeddingSessionTokenState = {
+  token: null,
+  loading: false,
+  error: null,
+};
+
+export const getSessionTokenState = (state: EnterpriseState) =>
+  state.embeddingSessionToken;
+
+const GET_OR_REFRESH_SESSION =
+  "metabase-enterprise/embeddingSessionToken/GET_OR_REFRESH_SESSION";
+
+export const getOrRefreshSession = createAsyncThunk(
+  GET_OR_REFRESH_SESSION,
+  async (url: string, { dispatch, getState }) => {
+    const state = getSessionTokenState(getState());
+    const token = state?.token;
+
+    const isTokenValid = token && token.exp * 1000 >= Date.now();
+
+    if (!state.loading && !isTokenValid) {
+      return dispatch(refreshTokenAsync(url));
+    }
+    return token;
+  },
+);
+const REFRESH_TOKEN = "metabase-enterprise/embeddingSessionToken/REFRESH_TOKEN";
+
+export const refreshTokenAsync = createAsyncThunk(
+  REFRESH_TOKEN,
+  async (url: string) => {
+    const response = await fetch(url, {
+      method: "GET",
+      credentials: "include",
+    });
+    return await response.json();
+  },
+);
+
+const tokenReducer = createReducer(initialState, builder =>
+  builder
+    .addCase(refreshTokenAsync.pending, state => {
+      state.loading = true;
+      return state;
```

Comment:
Do we need to return state here (and the other two places below)? I think not but just to make sure

Reply:
I thought it was actually an error, but apparently it's not.
We don't need to do it according to: https://immerjs.github.io/immer/return

> It is not needed to return anything from a producer, as Immer will return the (finalized) version of the draft anyway. However, it is allowed to just return draft.

But I think it's cleaner to not return anything, especially because there's another rule that says: ([from rtk](https://redux-toolkit.js.org/usage/immer-reducers#mutating-and-returning-state))

> In any given case reducer, Immer expects that you will either mutate the existing state, or construct a new state value yourself and return it, but not both in the same function!

Reply:
yeah, if we don't need to return anything, let's not return anything